Device event attributes

dNechita opened the pull request:
## PR Description
See discussion https://github.com/analogdevicesinc/libiio/pull/1368#issuecomment-3538607695.

Things left to do in this PR:

- [x] Update C#, C++ and Python bindings

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have commented new code, particularly complex or unclear areas
- [x] I have checked that I did not introduce new warnings or errors (CI output)
- [x] I have checked that components that use libiio did not get broken
- [ ] I have updated the documentation accordingly (GitHub Pages, READMEs, etc)

dNechita commented:
> I’m not convinced that the right abstraction here is something that simply exposes an interface where the user must poll for events. (it's better than what exists today, but still doesn't match what the normal use case is).
> 
> In the Linux IIO subsystem, events are conceptually hardware-driven notifications (threshold crossings, data ready, buffer watermarks, etc.), not just another value to periodically check. For example, the kernel exposes event channels that deliver timestamped events through a character device once they are enabled via sysfs.
> 
> Typical workflow (per the sysfs-bus-iio ABI):
> 
> * Enable the event through sysfs `/sys/bus/iio/devices/iio:deviceX/events/<event>_en`
> * Wait for the event on the device node (`/dev/iio:deviceX`)
> * Read the event structure once the file descriptor signals readiness.
> 
> This is inherently event-driven, not just attribute polling.
> 
> IMHO - what this means is we need a push model (`iiod` needs to push something to the connected clients) - not a pull model (where clients continually poll).

The push mode you are suggesting here would be a new feature and I agree that it would be a useful one. But I would create a new issue that requests such a feature, as this PR is not about the event propagation but about parsing iio attributes that happen to relate to events. To me, the below attributes indicate that a user could write values to them in order to configure how the events behave. E.g. how often some conditions to be checked (sampling_frequency) or what's the threshold above which an event should be issued (thresh_rising_value), etc.
├── events
│   ├── sampling_frequency
│   ├── sampling_frequency_available
│   ├── thresh_either_en
│   ├── thresh_falling_hysteresis
│   ├── thresh_falling_value
│   ├── thresh_rising_hysteresis
│   └── thresh_rising_value

Without the changes in this PR, the above attributes will be wrongly parsed.
These new attributes that will be exposed ar not for polling.

nunojsa commented:
So, conceptually I'm fine with the approach! We need to be able to enable events, hysteresis and so on! For the push stuff, conceptually it should have it's similarities with buffering! FWIW, I would assume we already support through:

https://github.com/analogdevicesinc/libiio/blob/main/events.c

Have you looked into the above? They need to properly integrate with each other

dNechita commented:
> So, conceptually I'm fine with the approach! We need to be able to enable events, hysteresis and so on! For the push stuff, conceptually it should have it's similarities with buffering! FWIW, I would assume we already support through:

It does have it's similarities with buffering but it's a polling strategy. Client code requests buffers at a time of its choosing. Same goes for the current implementation of events: client code creates a stream-event object and reads events whenever it sees fits. Events aren't lost as they pile up until they are read.

I see the push strategy as a way for client code to register a callback that would be executed at the time the event occurs. This is not implemented. Implementing this, would add quite a bit of complexity in the client - server architecture. As things are right now, IIOD always listens for client requests and only responds when asked. It never initiates anything. If we want the IIOD to take action when an event occurs, we need to setup a communication channel where the client code always listen for server notifications. This channel would exist simultaneously with the original communication channel.
> 
> https://github.com/analogdevicesinc/libiio/blob/main/events.c
> 
> Have you looked into the above? They need to properly integrate with each other

nunojsa commented:
> It does have it's similarities with buffering but it's a polling strategy. Client code requests buffers at a time of its choosing. Same goes for the current implementation of events: client code creates a stream-event object and reads events whenever it sees fits. Events aren't lost as they pile up until they are read.

And the above makes sense to me. If I'm not mistaken, events are very similar to buffers in the way they are implemented. Events are just pushed into a kernel fifo that you can read on /dev/iio....

> I see the push strategy as a way for client code to register a callback that would be executed at the time the event occurs. This is not implemented

Agreed this would be too much for now. You would need more thread handling (also on the client side) for taking care of the async nature of the above. For now I wouldn't complicate things. An app can always build that on top of libbio.

My 2 cents

dNechita commented:
V1:

- Amend the commit with the utils where I updated the code generation that is part of iio_attr to also handle the events attributes. In the same commit I fixed the iio_attr not showing event attributes in verbose mode.

nunojsa commented on emu.c:
```diff
@@ -732,6 +739,9 @@ static int check_available(const struct iio_attr *attr, const char *src)
 		break;
+	case IIO_ATTR_TYPE_EVENT:
+		avail_attr = iio_device_find_event_attr(attr->iio.dev, avail_name);
+		break;
```
Is the emu change really needed for this patch?

dNechita replied:
It can be in a different patch series to make it easier to review the core stuff first.

dNechita replied:
was moved into a separate commit

nunojsa commented on local.c:
```diff
@@ -1253,8 +1283,13 @@ static int add_attr_or_channel(void *d, const char *path)
+
+	if (is_event_attr(dev, name))
+		return iio_device_add_attr(dev, name, IIO_ATTR_TYPE_EVENT);
+	else
```
redundant else. 

But more important, this also looks a bit weird and misleading. Aren't all events attributes under events/? So `is_event_attr()` is a bit misleading given that the else has "events/".

I'm also not sure how we should handle this! As far as I can tell, static events attrs go into the device list and dynamic ones (which are defined as part of the channel definition) are part of the channel? I honestly think this is complicating things a bit but I see we already have this logic very deepen in that direction (same happens with scan_elements). But maybe there's room to improve things a bit. How will things look like if we do iio_info? We'll only see the "event attributes" for "device" kind of attributes right? The channel ones will come as "normal" channel attributes? 

So maybe we could introduce  the notion of event attributes for channels given that a channel can in  fact have an array of events:

https://elixir.bootlin.com/linux/v7.0.1/source/include/linux/iio/iio.h#L235

It could be even be a list. Then we could dump (and mark) channel event attributes when dumping channels. I think it would make things a bit more clear. But being objective, like this should work the same.

But up to you. Now that I think about it, it would also make sense to have APIs like iio_channel_read/write_event_attr(). Again, it would make things more clear about what is going on (even more given the fact these attrs can be enabled to be latter on buffered) but it would also complicate things a bit I guess.

dNechita replied:
You are right, channels will have attributes of type event and libiio should handle them.
I am thinking that adding `iio_channel_read/write_event_attr()` to the API is a good approach.

nunojsa commented on local.c:
```diff
@@ -1245,6 +1255,26 @@ static int add_attr_or_channel_helper(struct iio_device *dev,
+	}
+
+	/* If no channel prefix found, it's a device-level event attribute */
+	return true;
```
I think the API is a bit over commented

dNechita replied:
agree

nunojsa commented on iiod/lexer.l:
```diff
@@ -121,6 +121,11 @@ WORD (([[:alpha:]]+,)|(iio:))?(-|_|\.|[[:alnum:]])+
+<WANT_CHN_OR_ATTR>EVENT|event {
+	BEGIN(WANT_ATTR);
+	return EVENT_ATTR;
+}
```
I wonder if we need to care about this? This will be a libiio v1 only feature so do we care about the legacy protocol 😅?

dNechita replied:
In a realistic scenario this won't be needed since:
1. v1 client + v1 server use the binary protocol
2. v0 client + v1 server use ASCII protocol and v0 doesn't use have 'event attributes'

but
1. There is a very small chance that, during client–server negotiation, the client’s request to https://github.com/analogdevicesinc/libiio/blob/87f1497246ca1c573336fe8bce9cff5ac3c732da/iiod-client.c#L937 fails (for example, due to a sudden network drop causing a timeout). In that case, the connection falls back to using ASCII.
2. Alternatively, someone may be manually debugging via telnet or nc and sending READ device EVENT attr\r\n. This works because IIOD starts in ASCII mode by default and switches to binary mode only upon a client request.

I would keep it for now.

dNechita replied:
I took another look at it, and my arguments didn’t seem very compelling. I removed it.

nunojsa reviewed (commented):
Here it goes my 2 cents

gastmaier commented:
for the future new feature of not pooling, we would need, for different backends, having something similar to a (web)socket, where the connection is open and data is pushed to the host. the benefits are unclear.

I will test this pr tomorrow on hw

dNechita commented:
> I will test this pr tomorrow on hw

Thanks, your test results would be of great help.

Note that I will make additional changes to this PR to address the channel event attributes.

nunojsa commented:
> for the future new feature of not pooling, we would need, for different backends, having something similar to a (web)socket, where the connection is open and data is pushed to the host. the benefits are unclear.

I don't think we have pooling now in the "real" sense of the word. We just do `pool()` waiting for events. So it's kind of synchronous mechanism. An app will sleep/block waiting for events. The very same thing we have with buffers (unless I'm missing something). Of course that, one sees events and maybe the expectation is to get notified in an async way. But, honestly, I would defer that to the app rather than building such complexity in the library. At least for now...

dNechita commented:
V2:

- Introduced event-type channel attributes, previously only regular types existed.
- removed event attributes from the old ASCII parser
- updated emu backend in a separate commit 
- rebased on main

999Marius commented on context.c:
```diff
@@ -21,13 +21,14 @@ static const char xml_header[] = "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
 "<!ELEMENT context-attribute EMPTY>"
-"<!ELEMENT device (channel | buffer | attribute | debug-attribute | buffer-attribute)*>"
+"<!ELEMENT device (channel | buffer | attribute | debug-attribute | buffer-attribute | event-attribute)*>"
 "<!ELEMENT channel (scan-element?, attribute*)>"
```
Channel element should also include `event-attribute*` if not any XML with channel event attributes will fail when being parsed

999Marius commented on context.c:
```diff
@@ -38,6 +39,7 @@ static const char xml_header[] = "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
 "<!ATTLIST attribute name CDATA #REQUIRED filename CDATA #IMPLIED value CDATA #IMPLIED>"
 "<!ATTLIST debug-attribute name CDATA #REQUIRED value CDATA #IMPLIED>"
 "<!ATTLIST buffer-attribute name CDATA #REQUIRED value CDATA #IMPLIED>"
+"<!ATTLIST event-attribute name CDATA #REQUIRED value CDATA #IMPLIED>"
```
Here It seems we miss filename attribute `filename CDATA #IMPLIED` since `iio_snprintf_chan_event_attr_xml` adds a filename attribute when present

999Marius commented on device.c:
```diff
@@ -20,6 +20,7 @@ static const char * const xml_attr_prefix[] = {
 	"",
 	"debug-",
 	"buffer-",
+	"event-",
```
`"event-"`  should be between `buffer` and `debug`  since  `xml_attr_prefix[attr->type]` here would index by attr->type wich 
```
enum iio_attr_type {
	IIO_ATTR_TYPE_DEVICE = 0,
	IIO_ATTR_TYPE_DEBUG,
	IIO_ATTR_TYPE_EVENT,
	IIO_ATTR_TYPE_BUFFER,
	IIO_ATTR_TYPE_CHANNEL,
	IIO_ATTR_TYPE_CONTEXT,
};
```
and would put the prefix `buffer-` for `event-` and reverse by indexing with this enum

gastmaier replied:
Confirming they are indeed swapped
we can benefit from verbosity here
```
+ [IIO_ATTR_TYPE_DEVICE] = "",
+ [IIO_ATTR_TYPE_DEBUG]  = "debug-",
+ [IIO_ATTR_TYPE_EVENT]  = "event-",
+ [IIO_ATTR_TYPE_BUFFER] = "buffer-",
   ```

dNechita committed:
Add IIOD protocol support for event attributes

Extend the IIOD binary protocol to support reading and writing event
attributes for both devices and channels.

Add four new protocol opcodes:
- IIOD_OP_READ_DEV_EVT_ATTR: Read device event attributes
- IIOD_OP_WRITE_DEV_EVT_ATTR: Write device event attributes
- IIOD_OP_READ_CHN_EVT_ATTR: Read channel event attributes
- IIOD_OP_WRITE_CHN_EVT_ATTR: Write channel event attributes

Client-side changes (iiod-client.c):
- Add separate switch cases for IIO_ATTR_TYPE_DEVICE_EVENT and
  IIO_ATTR_TYPE_CHANNEL_EVENT in read/write operations
- For device event attributes: Pass device index and attribute index,
  use IIOD_OP_READ/WRITE_DEV_EVT_ATTR opcodes
- For channel event attributes: Pass device index, channel index, and
  attribute index, use IIOD_OP_READ/WRITE_CHN_EVT_ATTR opcodes
- Legacy text protocol returns -ENOSYS for event attributes (not
  supported in older protocol version)

Server-side changes (iiod-responder.h, iiod/responder.c):
- Define the four new opcodes in the protocol header
- Map opcodes to existing attribute read/write infrastructure in
  the responder

This enables remote access to event attributes through the IIOD
binary protocol.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

gastmaier reviewed (commented):
iio_info grouping works properly

```
root@analog-vck190:/sys/bus/iio/devices/iio:device0# echo 1000000 > events/sampling_frequency
root@analog-vck190:/sys/bus/iio/devices/iio:device0# echo 10 > events/thresh_falling_hysteresis
root@analog-vck190:/sys/bus/iio/devices/iio:device0# echo 5 > events/thresh_rising_value
root@analog-vck190:/sys/bus/iio/devices/iio:device0# iio_info
WARNING: Unable to read INI file: -5
iio_info version: 1.0 (git tag:af968f1)
Libiio version: 1.0 (git tag: af968f1) backends: local ip usb xml
IIO context created with local backend.
Backend version: 1.0 (git tag: af968f1)
Backend description string: Linux analog-vck190 6.12.0+ #9 SMP PREEMPT Mon May 11 19:23:09 CEST 2026 armv7l
IIO context has 2 attributes:
        attr  0: local,kernel value: 6.12.0+
        attr  1: uri value: local:
IIO context has 1 devices:
        iio:device0: ad4050 (buffer capable)
                1 channels found:
                        voltage0:  (input, index: 0, format: le:s12/32>>0)
                        6 channel-specific attributes found:
                                attr  0: calibscale (in_voltage_calibscale) value: 1.000000000
                                attr  1: oversampling_ratio (in_voltage_oversampling_ratio) value: 1
                                attr  2: oversampling_ratio_available (in_voltage_oversampling_ratio_available) options: 1 2 4 8 16 32 64 128 256
                                attr  3: raw (in_voltage_raw) value: 0
                                attr  4: sampling_frequency (in_voltage_sampling_frequency) value: 2000000
                                attr  5: scale (in_voltage_scale) value: 2.441406250
                2 buffer attributes found:
                        attr  0: data_available value: 0
                        attr  1: length_align_bytes value: 8
                1 debug attributes found:
                        attr  0: direct_reg_access
                7 device-specific event attributes found:
                        attr  0: sampling_frequency value: 1000000
                        attr  1: sampling_frequency_available options: 2000000 1000000 300000 100000 33300 10000 3000 500 333 250 200 166 140 124 111
                        attr  2: thresh_either_en value: 0
                        attr  3: thresh_falling_hysteresis value: 10
                        attr  4: thresh_falling_value value: 0
                        attr  5: thresh_rising_hysteresis value: 0
                        attr  6: thresh_rising_value value: 5
```
iio_attr missing, with https://github.com/analogdevicesinc/libiio/tree/device-event-attributes-iio-attr

```
iio_attr -d ad4050
iio_attr: Found ad4050 device, but it has 0 device attributes
dev 'ad4050', device event attr 'sampling_frequency', value '1000000'
dev 'ad4050', device event attr 'sampling_frequency_available', options '2000000' '1000000' '300000' '100000' '33300' '10000' '3000' '500' '333' '250' '200' '166' '140' '124' '111'
dev 'ad4050', device event attr 'thresh_either_en', value '0'
dev 'ad4050', device event attr 'thresh_falling_hysteresis', value '10'
dev 'ad4050', device event attr 'thresh_falling_value', value '0'
dev 'ad4050', device event attr 'thresh_rising_hysteresis', value '0'
dev 'ad4050', device event attr 'thresh_rising_value', value '5'
```

dNechita commented:
V3:

- reworked the commits as if both device and channel event attributes were added from the start
- added `-E` and `-e` flags to the `iio_attr` utility to handle device and channel attributes consistently with regular attributes
- rebased onto main, which includes recent multi-buffer changes, and resolved conflicts.

dNechita committed:
iio-backend: Add device and channel event types

Add two new attribute types to distinguish between device-level and
channel-level event attributes:

- IIO_ATTR_TYPE_DEVICE_EVENT: Device-level event attributes that apply
  to the entire device (e.g., calibration_complete_event_en)

- IIO_ATTR_TYPE_CHANNEL_EVENT: Channel-level event attributes that
  enable per-channel threshold monitoring (e.g.,
  in_accel_x_thresh_rising_en, in_voltage0_thresh_falling_value)

Both types are located in /sys/bus/iio/devices/iio:deviceX/events/ but
are distinguished by whether they contain channel identifiers.

Also add CHN_ATTRLIST_IDX() macro to convert channel attribute types
(IIO_ATTR_TYPE_CHANNEL, IIO_ATTR_TYPE_CHANNEL_EVENT) into array indices
for the channel's attrlist[] arrays. This macro relies on these types
being contiguous in the enum.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita committed:
treewide: Add storage infrastructure for event attributes

Extend device and channel structures to support separate storage for
device and channel event attributes:

- iio_device: Expand attrlist[] and values[] from 2 to 3 elements to
  store device, debug, and device_event attributes separately

- iio_channel: Expand attrlist[] and values[] from single to 2-element
  arrays to store channel and channel_event attributes separately,
  indexed via CHN_ATTRLIST_IDX() macro

Update attribute management functions:
- iio_device_add_attr(): Handle IIO_ATTR_TYPE_DEVICE_EVENT (index 2)
- iio_channel_add_attr(): Add type parameter to select target list,
  handle both IIO_ATTR_TYPE_CHANNEL and IIO_ATTR_TYPE_CHANNEL_EVENT

Update context lifecycle:
- Allocate/deallocate values arrays for all attribute types
- Free attributes from all lists during cleanup

Add backend detection support:
- local.c: Detect event attributes from sysfs events/ directory,
  distinguish device vs channel events by checking for channel ID
  prefixes (in_/out_), build sysfs paths for both types

- xml.c: Parse <event-attribute> tags, route to device or channel
  based on context

This commit adds the storage layer; subsequent commits will expose the
public API and add protocol/binding support.

Reject IIO_ATTR_TYPE_BUFFER in iio_device_add_attr(). Buffer attributes
have migrated to buffer ownership. Accepting this type would cause
out-of-bounds array access (attrlist[3] but max size is 2).

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita committed:
iio: Add event attribute API for devices and channels

Expose public API functions for accessing event attributes on both
devices and channels. These APIs enable applications to enumerate and
query event attributes that control hardware-driven threshold monitoring
without CPU polling.

Device event attribute API (added to include/iio/iio.h):
- iio_device_get_event_attrs_count() - Get number of device event attributes
- iio_device_get_event_attr() - Get device event attribute by index
- iio_device_find_event_attr() - Find device event attribute by name

Channel event attribute API (added to include/iio/iio.h):
- iio_channel_get_event_attrs_count() - Get number of channel event attributes
- iio_channel_get_event_attr() - Get channel event attribute by index
- iio_channel_find_event_attr() - Find channel event attribute by name

Add implementation in device.c and channel.c:
- Device functions access dev->attrlist[IIO_ATTR_TYPE_DEVICE_EVENT]
- Channel functions access chn->attrlist[CHN_ATTRLIST_IDX(IIO_ATTR_TYPE_CHANNEL_EVENT)]
- Channel XML serialization detects event attributes by type and uses
  <event-attribute> tags

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita committed:
utils: Add event attributes support to iio_info and iio_attr

Extend command-line utilities to display and manipulate event attributes
for both devices and channels.

iio_info changes:
- Display device event attributes after buffer attributes
- Display channel event attributes after regular channel attributes
- Format remains consistent with other attribute types

iio_attr changes:
- Add -E flag for device event attributes (uppercase for device-level,
  consistent with -D for debug)
- Add -e flag for channel event attributes (lowercase for channel-level,
  consistent with -c for channel)
- Syntax: -E [device] [attr] [value] for device event attributes
- Syntax: -e [device] [channel] [attr] [value] for channel event attributes
- Progressive discovery: flags alone show counts, with args access specific
  attributes
- Prevent ambiguity when channel names match attribute names by requiring
  explicit flags

Update man page (iio_attr.1.in):
- Document -E and -e flags
- Explain device vs channel event attribute access patterns

Event attributes enable efficient hardware monitoring in embedded systems:
device events for global state (calibration, faults), channel events for
per-channel thresholds (voltage limits, motion detection).

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita committed:
tests: Add comprehensive event attributes tests

Add automated tests for device and channel event attributes.

Changes:
- Add event_attrs.xml: Test device with device and channel event attributes
  - 1 device with device-level event attribute (events_enabled)
  - 3 channels with channel-level event attributes (threshold enables/values)

- Extend test_device.c: Add device_event_attributes() test function
  - Tests iio_device_get_event_attrs_count()
  - Tests iio_device_get_event_attr() with iteration
  - Tests iio_device_find_event_attr() by name
  - Tests invalid index and nonexistent attribute handling

  - Extend test_channel.c: Add channel_event_attributes() test function
  - Tests iio_channel_get_event_attrs_count()
  - Tests iio_channel_get_event_attr() with iteration
  - Tests iio_channel_find_event_attr() by name
  - Tests attribute name and filename retrieval
  - Tests invalid index and nonexistent attribute handling

Tests pass gracefully on backends without event attributes.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita committed:
bindings: Add support for event attributes

Extend Python, C#, and C++ bindings to expose event attributes for
both devices and channels.

Python bindings (bindings/python/iio/__init__.py):
- Add ctypes function bindings for iio_device_get_event_attrs_count(),
  iio_device_get_event_attr(), iio_channel_get_event_attrs_count(),
  and iio_channel_get_event_attr()
- Add event_attrs property to Device class returning dictionary of
  device event attributes
- Add event_attrs property to Channel class returning dictionary of
  channel event attributes

C# bindings (bindings/csharp/):
- Add DllImport declarations for event attribute functions in Device.cs
  and Channel.cs
- Add event_attrs readonly List<Attr> to Device and Channel classes
- Initialize and populate event_attrs lists in Device and Channel
  constructors using the C API functions

C++ bindings (bindings/cpp/iio.hpp):
- Add EventAttrSeq typedef for both Device and Channel classes using
  the AttrSeqT template
- Add event_attrs member to Device and Channel, initialized in
  constructors
- Add event_attrs_count(), event_attr(), and find_event_attr() methods
  to both Device and Channel classes

The bindings follow the same pattern as existing debug attributes,
providing consistent access to event attributes across all language
bindings.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita committed:
Add EMU backend support for event attributes

Extend the EMU (emulation) backend to support reading and writing event
attributes for both devices and channels using proper type-based
detection.

Device event attributes:
- Add IIO_ATTR_TYPE_DEVICE_EVENT case to getDeviceAttr() switch for
  XML attribute lookup
- Parse <event-attribute> tags in create_device() and add them using
  IIO_ATTR_TYPE_DEVICE_EVENT type
- Add IIO_ATTR_TYPE_DEVICE_EVENT case to check_available() for
  validating values against _available attributes

Channel event attributes:
- Add getChannelEventAttr() function to locate channel event attributes
  in XML by searching for <event-attribute> tags
- Add read_channel_event_attr() and write_channel_event_attr() functions
  to handle XML read/write operations
- Add add_event_attr_to_channel() to parse <event-attribute> XML nodes
  and add them using IIO_ATTR_TYPE_CHANNEL_EVENT type
- Parse <event-attribute> tags in create_channel()

Runtime attribute access:
- Add IIO_ATTR_TYPE_CHANNEL_EVENT case to emu_read_attr() to route
  channel event attribute reads to read_channel_event_attr()
- Add IIO_ATTR_TYPE_CHANNEL_EVENT case to emu_write_attr() to route
  channel event attribute writes to write_channel_event_attr()
- Add IIO_ATTR_TYPE_CHANNEL_EVENT case to check_available() to find
  and validate against _available attributes for channel events

The EMU backend now properly distinguishes event attributes using the
IIO_ATTR_TYPE_DEVICE_EVENT and IIO_ATTR_TYPE_CHANNEL_EVENT types,
allowing XML-based emulation of devices with event attributes.

Signed-off-by: Dan Nechita <dan.nechita@analog.com>

dNechita commented:
V4:

- The initial approach introduced a new attribute type, `IIO_ATTR_TYPE_EVENT`. However, this came with limitations, as there was no clear way to distinguish between channel event attributes and device event attributes. As a result, workarounds were introduced, such as relying on the attribute’s filename structure and checking for an "events" suffix. o address this, I introduced `IIO_ATTR_TYPE_DEVICE_EVENT` and `IIO_ATTR_TYPE_CHANNEL_EVENT`, which resolve the issue and provide a much cleaner separation.
- The above change impacted other parts of the implementation, but it did not introduce any functional differences. The utils and bindings remain unchanged.
- Updated the API tests to cover the newly introduced API functions

dNechita commented:
@gastmaier @nunojsa @999Marius I am ready to merge this. It would be great if you could take another look at this.

dNechita commented:
V5:

- fix in `iio_device_add_attr()`: buffer-type attributes are now rejected, preventing out-of-bounds access to `dev->attrlist[3]` when the maximum size is 2.

999Marius reviewed (approved):
LGTM

gastmaier reviewed (commented):
Tested per previous comment, values are correctly set:

```
$ iio_info
IIO context has 2 attributes:
        attr  0: local,kernel value: 6.12.0+
        attr  1: uri value: local:
IIO context has 1 devices:
        iio:device0: ad4050 (buffer capable)
                1 channels found:
                        voltage0:  (input, index: 0, format: le:s14/32>>0)
                        6 channel-specific attributes found:
                                attr  0: calibscale (in_voltage_calibscale) value: 1.000000000
                                attr  1: oversampling_ratio (in_voltage_oversampling_ratio) value: 4
                                attr  2: oversampling_ratio_available (in_voltage_oversampling_ratio_available) options: 1 2 4 8 16 32 64 128 256
                                attr  3: raw (in_voltage_raw) value: -516
                                attr  4: sampling_frequency (in_voltage_sampling_frequency) value: 2000000
                                attr  5: scale (in_voltage_scale) value: 0.610351562
                1 buffer found:
                        buffer0:  (input)
                        1 scan elements found:
                                Scan element 0: voltage0 (input)
                        2 buffer attributes found:
                                attr  0: data_available value: 1600
                                attr  1: length_align_bytes value: 8
                7 event attributes found:
                        attr  0: sampling_frequency value: 1000000
                        attr  1: sampling_frequency_available options: 2000000 1000000 300000 100000 33300 10000 3000 500 333 250 200 166 140 124 111
                        attr  2: thresh_either_en value: 0
                        attr  3: thresh_falling_hysteresis value: 10
                        attr  4: thresh_falling_value value: 0
                        attr  5: thresh_rising_hysteresis value: 0
                        attr  6: thresh_rising_value value: 5
                1 debug attributes found:
                        attr  0: direct_reg_access
                No trigger on this device
```

dNechita commented:
Awesome. Thanks for testing this!
